### PR TITLE
(React SDK): Forward refs in withOptimizely HoC

### PR DESCRIPTION
**Pull request by mjc1283**
Implement ref forwarding in `withOptimizely` HoC.

### CODE REVIEW

File: packages/react-sdk/src/withOptimizely.tsx [hunk 1]
```diff
@@ -17,36 +17,44 @@ import * as React from 'react'
...
+            optimizely={value.optimizely}
+            optimizelyReadyTimeout={value.timeout}
+            isServerSide={value.isServerSide}
+            ref={ref}
+          />
+        )}
+      </OptimizelyContextConsumer>
+    )
   }
+  withOptimizely.displayName = `withOptimizely(${Component.displayName ||
```
Comment (jamesopti) [inline packages/react-sdk/src/withOptimizely.tsx]: Checkout the helper I wrote in the segment-js repo to forward refs and hoist non-react statics.

I think copy/pasting that here would be useful.

https://github.com/optimizely/segment-js/blob/master/src/decorators.js#L14-L29

File: packages/react-sdk/src/withOptimizely.spec.tsx [hunk 2]
```diff
@@ -192,4 +192,35 @@ describe('withOptimizely', () => {
...
+        timeout={200}
+        user={{ id: 'jordan' }}
+        userAttributes={{ plan_type: 'bronze' }}
+        isServerSide={true}
+      >
+        <OptimizelyInput ref={inputRef} defaultValue="hi" />
+      </OptimizelyProvider>,
+    )
+    expect(inputRef.current).toBeInstanceOf(HTMLInputElement)
+    expect(typeof inputRef.current!.focus).toBe('function')
```
Comment (jamesopti) [inline packages/react-sdk/src/withOptimizely.spec.tsx]: Can you also assert that `inputRef.current.value` is in fact the same reference as `component.find('input'). getDOMNode()`?

File: packages/react-sdk/src/withOptimizely.tsx (hunk 1 above)
Comment (mjc1283) [reply packages/react-sdk/src/withOptimizely.tsx]: @jamesopti I tried using this helper function, but I'm finding this to make the types a lot more challenging. Would you be OK if we defer this until later? FWIW, this `withOptimizely` is the only HoC that would be using this helper function at the moment.

**Comment by mjc1283** [comment]
@jamesopti @mikeng13 I refactored this a little and combined it with hoisting statics. Moved into a separate PR: https://github.com/optimizely/fullstack-labs/pull/32

**Pull request closed without merging**